### PR TITLE
Default authZ server with role bindings API and provider integration

### PR DESCRIPTION
**Pull request by aayustark007-fk**
- [x] Role Bindings Update API
- [ ] AuthZ for update API for Role Bindings (need help)
- [x] Calling AuthZ server from Default provider
- [x] Tests for AuthZ server
- [x] Tests for integration of provider with server

### CODE REVIEW

File: authz/build.gradle [hunk 1]
```diff
@@ -3,15 +3,14 @@ plugins {
     id 'java-test-fixtures'
 }
 
-dependencies {
-    implementation(project(':common'))
+components.java.withVariantsFromConfiguration(configurations.testFixturesApiElements) {
+    skip()
```
Comment (aayustark007-fk) [inline authz/build.gradle:7]: required to prevent publishing test-fixtures artefacts

File: authz/src/main/java/com/flipkart/varadhi/authz/AuthorizationProvider.java [hunk 2]
```diff
@@ -1,19 +1,20 @@
...
-import com.flipkart.varadhi.entities.ResourceAction;
 import com.flipkart.varadhi.config.AuthorizationOptions;
+import com.flipkart.varadhi.entities.ResourceAction;
 import com.flipkart.varadhi.entities.UserContext;
 import io.vertx.core.Future;
+import io.vertx.core.Vertx;
 
 public interface AuthorizationProvider {
-    Future<Boolean> init(AuthorizationOptions authorizationOptions);
+    Future<Boolean> init(Vertx vertx, AuthorizationOptions authorizationOptions);
```
Comment (aayustark007-fk) [inline authz/src/main/java/com/flipkart/varadhi/authz/AuthorizationProvider.java:10]: required for webclient

Comment (aayustark007-fk) [reply authz/src/main/java/com/flipkart/varadhi/authz/AuthorizationProvider.java:10]: we can use `Vertx.vertx()` but can that cause issues?

File: common/build.gradle [hunk 3]
```diff
@@ -1,5 +1,5 @@
 plugins {
-    id 'com.flipkart.varadhi.java-published-library-conventions'
+    id 'com.flipkart.varadhi.java-library-conventions'
```
Comment (aayustark007-fk) [inline common/build.gradle:2]: authz does not depend on common anymore, so no need to publish this

File: entities/src/main/java/com/flipkart/varadhi/auth/RoleBindingNode.java [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+    ResourceType resourceType;
+    Map<String, Set<String>> subjectToRolesMapping;
+
+    public RoleBindingNode(
+            String resourceId, ResourceType resourceType, Map<String, Set<String>> subjectToRolesMapping, int version
+    ) {
+        super(resourceId, version);
+        this.resourceId = resourceId;
+        this.resourceType = resourceType;
+        this.subjectToRolesMapping = new HashMap<>();
```
Comment (aayustark007-fk) [inline entities/src/main/java/com/flipkart/varadhi/auth/RoleBindingNode.java:27]: need to ensure that `subjectsToRoleMapping` is modifiable map. It is possible to instantiate by passing unmodifiable map via `Map.of()` which will cause the `setRoleAssignment` method to throw unmodifiable collection exception

File: server/src/main/java/com/flipkart/varadhi/VerticleDeployer.java [hunk 5]
```diff
@@ -104,10 +109,46 @@ private List<RouteDefinition> getDefinitions() {
...
+                .collect(Collectors.toList());
+    }
+
 
     public void deployVerticle(
             Vertx vertx,
             ServerConfiguration configuration
+    ) {
+        if (configuration.isDefaultAuthorizationServerEnabled()) {
+            deployDefaultAuthZVerticle(vertx, configuration);
```
Comment (aayustark007-fk) [inline server/src/main/java/com/flipkart/varadhi/VerticleDeployer.java:127]: deploying as a separate verticle will not cause issues with roundrobin

Comment (aayustark007-fk) [reply server/src/main/java/com/flipkart/varadhi/VerticleDeployer.java:127]: have tested locally

File: server/src/main/java/com/flipkart/varadhi/services/DefaultAuthZService.java [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+import com.flipkart.varadhi.entities.ResourceType;
+import com.flipkart.varadhi.entities.RoleAssignmentUpdate;
+import com.flipkart.varadhi.exceptions.InvalidOperationForResourceException;
+import com.flipkart.varadhi.exceptions.ResourceNotFoundException;
+import com.flipkart.varadhi.spi.db.MetaStore;
+
+import java.util.HashMap;
+import java.util.List;
+
+public class DefaultAuthZService {
```
Comment (kmrdhruv) [inline server/src/main/java/com/flipkart/varadhi/services/DefaultAuthZService.java:13]: AuthzService or iamService or something equivalent ?

File: server/src/main/java/com/flipkart/varadhi/services/DefaultAuthZService.java [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+                    resourceId,
+                    resourceType
+            ));
+        }
+        RoleBindingNode node = new RoleBindingNode(resourceId, resourceType, new HashMap<>(), 0);
+        metaStore.createRoleBindingNode(node);
+        return node;
+    }
+
+    public List<RoleBindingNode> getAllRoleBindingNodes() {
```
Comment (kmrdhruv) [inline server/src/main/java/com/flipkart/varadhi/services/DefaultAuthZService.java:33]: Why is this needed ? I mean any use cases which needs all role binding nodes ?

File: server/src/main/java/com/flipkart/varadhi/services/DefaultAuthZService.java [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+        return metaStore.getRoleBindingNodes();
+    }
+
+    public RoleBindingNode getRoleBindingNode(String resourceId) {
```
Comment (kmrdhruv) [inline server/src/main/java/com/flipkart/varadhi/services/DefaultAuthZService.java:37]: would iampolicy better instead of RoleBindingNode ?

File: server/src/main/java/com/flipkart/varadhi/services/DefaultAuthZService.java [hunk 9]
```diff
@@ -0,0 +1,119 @@
...
+    }
+
+    private RoleBindingNode createOrGetRoleBindingNode(String resourceId, ResourceType resourceType) {
+        boolean exists = metaStore.checkRoleBindingNodeExists(resourceId);
+        if (!exists) {
+            return createRoleBindingNode(resourceId, resourceType);
+        }
+        RoleBindingNode existingNode = metaStore.getRoleBindingNode(resourceId);
+        if (existingNode.getResourceType() != resourceType) {
+            throw new InvalidOperationForResourceException(String.format(
```
Comment (kmrdhruv) [inline server/src/main/java/com/flipkart/varadhi/services/DefaultAuthZService.java:88]: invalidArgument ?

File: server/src/test/java/com/flipkart/varadhi/auth/DefaultAuthorizationProviderTest.java [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+        defaultAuthorizationProvider.init(Vertx.vertx(), authorizationOptions)
+                .onComplete(testContext.succeeding(t -> {
+                    Assertions.assertTrue(t);
+                    checkpoint.flag();
+                }));
+    }
+
+    @Test
+    public void testNotInit() {
+        Assertions.assertThrows(InvalidConfigException.class, () ->
```
Comment (kmrdhruv) [inline server/src/test/java/com/flipkart/varadhi/auth/DefaultAuthorizationProviderTest.java:89]: would it be invalidConfig for not initialized  ?

**Review by kmrdhruv — CHANGES_REQUESTED** [review]

File: authz/build.gradle (hunk 1 above)
Comment (aayustark007-fk) [reply authz/build.gradle:7]: when I add this to conventions, the build starts failing as it is not able to recognise the  `testFixturesApiElements` configurations

File: entities/src/main/java/com/flipkart/varadhi/auth/RoleBindingNode.java [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+    Map<String, Set<String>> subjectToRolesMapping;
+
+    @JsonCreator
+    public RoleBindingNode(
+            @JsonProperty("resourceId") String resourceId,
+            @JsonProperty("resourceType") ResourceType resourceType,
+            @JsonProperty("subjectToRolesMapping")
+            Map<String, Set<String>> subjectToRolesMapping,
+            @JsonProperty("version")
+            int version
```
[the remark this answers is not in the document]
Comment (aayustark007-fk) [reply entities/src/main/java/com/flipkart/varadhi/auth/RoleBindingNode.java:30]: E2E tests were failing with could not construct instance error. Hence, the json creator annotation

File: server/src/main/java/com/flipkart/varadhi/VerticleDeployer.java (hunk 5 above)
Comment (aayustark007-fk) [reply server/src/main/java/com/flipkart/varadhi/VerticleDeployer.java:127]: mainly thinking from pov of isolation
if not needed then we can have it as part of single router

File: server/src/main/java/com/flipkart/varadhi/auth/DefaultAuthorizationProvider.java [hunk 12]
```diff
@@ -76,38 +87,66 @@ private List<Pair<ResourceType, String>> resolveOrderedFromLeaf(ResourceAction a
...
     }
 
-    private Set<String> getRolesForSubject(String subject, String resourceId) {
-        return configuration.getRoleBindings()
-                .getOrDefault(resourceId, Map.of())
-                .getOrDefault(subject, Set.of());
+
+    protected Future<Set<String>> getRolesForSubject(String subject, String resourceId) {
+        // use web client to make get call
+        return webClient.get(
```
[the remark this answers is not in the document]
Comment (aayustark007-fk) [reply server/src/main/java/com/flipkart/varadhi/auth/DefaultAuthorizationProvider.java:114]: config is to set the method and response code to cache
by default it is GET methods and 200, 301 and 404 codes are cached
response is cached as HttpResponse which needs to be decoded

File: server/src/main/java/com/flipkart/varadhi/config/DefaultAuthorizationConfiguration.java [hunk 13]
```diff
@@ -0,0 +1,18 @@
...
+@Data
+public class DefaultAuthorizationConfiguration {
+    /**
+     * role_id to Role{role_id, [permissions...]} mappings
+     */
+    private Map<String, Role> roleDefinitions;
+
+    private String authZServerHost = "localhost";
+    private int authZServerPort = 8088;
+    private String authZServerPath = "/v1/authz/bindings/";
```
[the remark this answers is not in the document]
Comment (aayustark007-fk) [reply server/src/main/java/com/flipkart/varadhi/config/DefaultAuthorizationConfiguration.java:17]: I added it to prevent hardcoding of values but also giving default values so that someone can directly run container without having to configure it

File: server/src/main/java/com/flipkart/varadhi/db/VaradhiMetaStore.java [hunk 14]
```diff
@@ -219,4 +221,39 @@ public void deleteVaradhiTopic(String varadhiTopicName) {
...
     }
 
+    @Override
+    public List<RoleBindingNode> getRoleBindingNodes() {
+        ZNode znode = ZNode.OfEntityType(ROLE_BINDING);
+        return zkMetaStore.listChildren(znode).stream().map(this::getRoleBindingNode).toList();
+    }
+
+    @Override
+    public RoleBindingNode getRoleBindingNode(String resourceId) {
```
[the remark this answers is not in the document]
Comment (aayustark007-fk) [reply server/src/main/java/com/flipkart/varadhi/db/VaradhiMetaStore.java:231]: yes that can cause issues, especially at project level for same project name under different orgs.
But, won't the same issue happen with varadhi entities too?

File: server/src/main/java/com/flipkart/varadhi/auth/DefaultAuthorizationProvider.java [hunk 15]
```diff
@@ -50,9 +58,12 @@ public Future<Boolean> isAuthorized(UserContext userContext, ResourceAction acti
             throw new InvalidConfigException("Default Authorization Provider is not initialised.");
         }
         List<Pair<ResourceType, String>> leafToRootResourceIds = resolveOrderedFromLeaf(action, resource);
-        var result = leafToRootResourceIds.stream().filter(pair -> StringUtils.isNotBlank(pair.getValue()))
-                .anyMatch(pair -> isAuthorizedInternal(userContext.getSubject(), action, pair.getValue()));
-        return Future.succeededFuture(result);
+        List<Future<Boolean>> futures =
+                leafToRootResourceIds.stream().filter(pair -> StringUtils.isNotBlank(pair.getValue()))
```
[the remark this answers is not in the document]
Comment (aayustark007-fk) [reply server/src/main/java/com/flipkart/varadhi/auth/DefaultAuthorizationProvider.java:62]: adding case to handle empty list, where it should return false (currently it will cause it to return true)

File: server/src/main/java/com/flipkart/varadhi/services/DefaultAuthZService.java (hunk 7 above)
Comment (aayustark007-fk) [reply server/src/main/java/com/flipkart/varadhi/services/DefaultAuthZService.java:33]: this is mainly for debug purposes

File: common/src/main/java/com/flipkart/varadhi/exceptions/NotInitializedException.java [hunk 16]
```diff
@@ -0,0 +1,7 @@
+package com.flipkart.varadhi.exceptions;
+
+import lombok.experimental.StandardException;
+
+@StandardException
+public class NotInitializedException extends VaradhiException {
```
Comment (gauravAshok) [inline common/src/main/java/com/flipkart/varadhi/exceptions/NotInitializedException.java:6]: No need to have this. This type of error falls in IllegalStateExceptions bucket. Something was expected but was not true. It will happen due to a bug and not a common occurrence.

File: core/build.gradle [hunk 17]
```diff
@@ -6,6 +6,7 @@ plugins {
 dependencies {
     implementation(project(':common'))
     implementation(project(':entities'))
+    implementation(project(':authz'))
```
[the remark this answers is not in the document]
Comment (gauravAshok) [reply core/build.gradle:9]: Ya. no module will depend on it in gradle.

File: core/src/main/java/com/flipkart/varadhi/spi/db/MetaStore.java [hunk 18]
```diff
@@ -1,5 +1,6 @@
 package com.flipkart.varadhi.spi.db;
 
+import com.flipkart.varadhi.auth.RoleBindingNode;
```
[the remark this answers is not in the document]
Comment (gauravAshok) [reply core/src/main/java/com/flipkart/varadhi/spi/db/MetaStore.java:3]: +1. separate interface.

File: entities/build.gradle [hunk 19]
```diff
@@ -3,6 +3,7 @@ plugins {
 }
 dependencies {
     implementation(project(":common"))
+    implementation(project(":authz"))
```
Comment (gauravAshok) [inline entities/build.gradle:6]: you can remove it.

File: entities/src/main/java/com/flipkart/varadhi/auth/RoleBindingNode.java [hunk 20]
```diff
@@ -0,0 +1,46 @@
...
+
+@Getter
+@EqualsAndHashCode(callSuper = true)
+@ValidateVaradhiResource(message = "Invalid Role Binding name. Check naming constraints.")
+public class RoleBindingNode extends VaradhiResource {
+    String resourceId;
+    ResourceType resourceType;
+    Map<String, Set<String>> subjectToRolesMapping;
+
+    @JsonCreator
```
Comment (gauravAshok) [inline entities/src/main/java/com/flipkart/varadhi/auth/RoleBindingNode.java:23]: With json creator, the class fields can be final.

File: core/build.gradle (hunk 17 above)
[the remark this answers is not in the document]
Comment (AayuStark007) [reply core/build.gradle:9]: this is due to the usage of ResourceType enum which belongs to authz module

File: core/src/main/java/com/flipkart/varadhi/spi/db/MetaStore.java (hunk 18 above)
[the remark this answers is not in the document]
Comment (AayuStark007) [reply core/src/main/java/com/flipkart/varadhi/spi/db/MetaStore.java:3]: I tried separating this out but it will lead to introduction of zookeeper dependencies in VerticleDeployer.
Otherwise, we will have to follow the metastore provider pattern for AuthZ too which will involve quite a lot of changes.

File: entities/build.gradle (hunk 19 above)
Comment (AayuStark007) [reply entities/build.gradle:6]: this is due to dependency on ResourceType which is used in RoleBindingNode entity

File: core/build.gradle (hunk 17 above)
[the remark this answers is not in the document]
Comment (aayustark007-fk) [reply core/build.gradle:9]: this is due to the usage of ResourceType enum which belongs to authz module

File: core/src/main/java/com/flipkart/varadhi/spi/db/MetaStore.java (hunk 18 above)
[the remark this answers is not in the document]
Comment (aayustark007-fk) [reply core/src/main/java/com/flipkart/varadhi/spi/db/MetaStore.java:3]: I tried separating this out but it will lead to introduction of zookeeper dependencies in VerticleDeployer.
Otherwise, we will have to follow the metastore provider pattern for AuthZ too which will involve quite a lot of changes.

File: entities/build.gradle (hunk 19 above)
Comment (aayustark007-fk) [reply entities/build.gradle:6]: this is due to dependency on ResourceType which is used in RoleBindingNode entity

File: server/src/main/java/com/flipkart/varadhi/web/v1/authz/AuthZHandlers.java [hunk 21]
```diff
@@ -0,0 +1,205 @@
...
+                                Set.of(authenticated),
+                                new LinkedHashSet<>(),
+                                this.getIAMPolicyHandler(ResourceType.ORG),
+                                true,
+                                Optional.empty()
+                        ),
+                        new RouteDefinition(
+                                HttpMethod.PUT,
+                                "/orgs/:org/policy",
+                                Set.of(hasBody, authenticated),
```
Comment (gauravAshok) [inline server/src/main/java/com/flipkart/varadhi/web/v1/authz/AuthZHandlers.java:88]: Since there are repetitions here,
we can put common patterns in the RouteDefinition itself  as static methods.

```java
public static Builder get(String path) {
        return new Builder(HttpMethod.GET, path);
    }

    public static Builder put(String path) {
        return new Builder(HttpMethod.PUT, path);
    }

    public static Builder post(String path) {
        return new Builder(HttpMethod.POST, path);
    }

    public static Builder delete(String path) {
        return new Builder(HttpMethod.DELETE, path);
    }

    @RequiredArgsConstructor
    public static class Builder {

        private final HttpMethod method;

        private final String path;

        private boolean authenticated;

        private boolean hasBody;

        private boolean blocking;

        public Builder authenticated() {
            this.authenticated = true;
            return this;
        }

        public Builder hasBody() {
            this.hasBody = true;
            return this;
        }

        public Builder blocking() {
            this.blocking = true;
            return this;
        }

        public RouteDefinition build(Handler<RoutingContext> reqHandler) {
            Set<RouteBehaviour> behaviours = new LinkedHashSet<>();
            if (authenticated) {
                behaviours.add(RouteBehaviour.authenticated);
            }
            if (hasBody) {
                behaviours.add(RouteBehaviour.hasBody);
            }
            return new RouteDefinition(
                    method,
                    path,
                    behaviours,
                    new LinkedHashSet<>(),
                    reqHandler,
                    blocking,
                    Optional.empty()
            );
        }
    }
```

Then these usages will look like
```java
RouteDefinition.get("/orgs/:org/policy")
                                .authenticated()
                                .blocking()
                                .build(this.getIAMPolicyHandler(ResourceType.ORG)),
```

File: server/src/main/java/com/flipkart/varadhi/web/v1/authz/AuthZHandlers.java [hunk 22]
```diff
@@ -0,0 +1,205 @@
...
+    public List<RouteDefinition> get() {
+        return Stream.of(
+                getDebugHandlers(),
+                getPolicyHandlers()
+        ).flatMap(List::stream).toList();
+    }
+
+    public Handler<RoutingContext> getIAMPolicyHandler(ResourceType resourceType) {
+        return (routingContext) -> {
+            String resourceId = routingContext.getResourceIdFromPath(resourceType);
```
Comment (gauravAshok) [inline server/src/main/java/com/flipkart/varadhi/web/v1/authz/AuthZHandlers.java:162]: Putting getResourceIdFromPath in extension is a code smell.
This method should lie here as a normal method.

Extension methods are pure utility methods. They dont handle business logic.

File: server/src/main/java/com/flipkart/varadhi/web/v1/authz/AuthZHandlers.java [hunk 23]
```diff
@@ -0,0 +1,205 @@
...
+    }
+
+    public RoleBindingNode getIAMPolicy(ResourceType resourceType, String resourceId) {
+        return authZService.getIAMPolicy(resourceType, resourceId);
+    }
+
+    public Handler<RoutingContext> setIAMPolicyHandler(ResourceType resourceType) {
+        return (routingContext) -> {
+            String resourceId = routingContext.getResourceIdFromPath(resourceType);
+            IAMPolicyRequest policyForSubject = routingContext.body().asPojo(IAMPolicyRequest.class);
```
Comment (gauravAshok) [inline server/src/main/java/com/flipkart/varadhi/web/v1/authz/AuthZHandlers.java:175]: IAMPolicyRequest looks like can go to entities module.
entities.auth package.

File: server/src/main/java/com/flipkart/varadhi/web/Extensions.java [hunk 24]
```diff
@@ -98,5 +102,19 @@ public static <T> T getApiResponse(RoutingContext ctx) {
         public static void todo(RoutingContext context) {
             throw new NotImplementedException("Not Implemented.");
         }
+
+        public static String getResourceIdFromPath(RoutingContext ctx, ResourceType resourceType) {
```
Comment (gauravAshok) [inline server/src/main/java/com/flipkart/varadhi/web/Extensions.java:106]: as suggested elsewhere, this method can be moved to AuthzHandler only.

File: server/src/main/java/com/flipkart/varadhi/config/ServerConfiguration.java [hunk 25]
```diff
@@ -28,6 +28,7 @@ public class ServerConfiguration {
 
     private AuthorizationOptions authorization;
 
+    private boolean authorizationServerEnabled;
```
Comment (gauravAshok) [inline server/src/main/java/com/flipkart/varadhi/config/ServerConfiguration.java:31]: can we make it implicit.
Can server be disabled, if I configure DefaultAuthorizationProvider as the authzProvider?

File: server/src/test/java/com/flipkart/varadhi/auth/DefaultAuthorizationProviderTest.java [hunk 26]
```diff
@@ -0,0 +1,490 @@
...
+
+    private DefaultAuthorizationProvider provider;
+
+    private AuthZService authZService;
+
+    @BeforeEach
+    public void setUp() throws IOException {
+        String configContent =
+                """
+                        ---
```
Comment (gauravAshok) [inline server/src/test/java/com/flipkart/varadhi/auth/DefaultAuthorizationProviderTest.java:44]: should we put this in file?

**Review by gauravAshok — CHANGES_REQUESTED** [review]

**Comment by gauravAshok** [comment]
The authz handlers are marked authenticated, but they dont have authz checks. Based on your description, anything pending there?

File: entities/src/main/java/com/flipkart/varadhi/entities/auth/RoleBindingNode.java [hunk 27]
```diff
@@ -0,0 +1,52 @@
...
+    private final ResourceType resourceType;
+
+    @NotNull
+    private final Map<String, Set<String>> subjectToRolesMapping;
+
+    @JsonCreator
+    public RoleBindingNode(
+            @JsonProperty("resourceId") String resourceId,
+            @JsonProperty("resourceType") ResourceType resourceType,
+            @JsonProperty("subjectToRolesMapping")
```
[the remark this answers is not in the document]
Comment (aayustark007-fk) [reply entities/src/main/java/com/flipkart/varadhi/entities/auth/RoleBindingNode.java:33]: this to indicate that the map keys are subject names, otherwise there can be confusion whether its subject to roles mapping or vice versa

File: authz/src/testFixtures/java/com/flipkart/varadhi/entities/TestUser.java [hunk 28]
```diff
@@ -1,5 +1,7 @@
 package com.flipkart.varadhi.entities;
```
[the remark this answers is not in the document]
Comment (aayustark007-fk) [reply authz/src/testFixtures/java/com/flipkart/varadhi/entities/TestUser.java:1]: done

File: authz/src/main/java/com/flipkart/varadhi/authz/AuthorizationProvider.java [hunk 29]
```diff
@@ -1,8 +1,8 @@
 package com.flipkart.varadhi.authz;
 
-import com.flipkart.varadhi.entities.ResourceAction;
 import com.flipkart.varadhi.config.AuthorizationOptions;
-import com.flipkart.varadhi.entities.UserContext;
+import com.flipkart.varadhi.entities.auth.ResourceAction;
+import com.flipkart.varadhi.entities.auth.UserContext;
```
[the remark this answers is not in the document]
Comment (aayustark007-fk) [reply authz/src/main/java/com/flipkart/varadhi/authz/AuthorizationProvider.java:5]: done

File: server/src/main/java/com/flipkart/varadhi/web/v1/authz/AuthZHandlers.java [hunk 30]
```diff
@@ -0,0 +1,205 @@
...
+                                new LinkedHashSet<>(),
+                                this::deleteRoleBindingNode,
+                                true,
+                                Optional.empty()
+                        )
+                )
+        ).get();
+    }
+
+    public List<RouteDefinition> getPolicyHandlers() {
```
[the remark this answers is not in the document]
Comment (aayustark007-fk) [reply server/src/main/java/com/flipkart/varadhi/web/v1/authz/AuthZHandlers.java:72]: done

File: server/src/main/java/com/flipkart/varadhi/web/v1/authz/AuthZHandlers.java (hunk 21 above)
Comment (aayustark007-fk) [reply server/src/main/java/com/flipkart/varadhi/web/v1/authz/AuthZHandlers.java:88]: done

File: server/src/main/java/com/flipkart/varadhi/web/v1/authz/AuthZHandlers.java (hunk 22 above)
Comment (aayustark007-fk) [reply server/src/main/java/com/flipkart/varadhi/web/v1/authz/AuthZHandlers.java:162]: done

File: server/src/main/java/com/flipkart/varadhi/web/v1/authz/AuthZHandlers.java (hunk 23 above)
Comment (aayustark007-fk) [reply server/src/main/java/com/flipkart/varadhi/web/v1/authz/AuthZHandlers.java:175]: done

File: server/src/main/java/com/flipkart/varadhi/web/Extensions.java (hunk 24 above)
Comment (aayustark007-fk) [reply server/src/main/java/com/flipkart/varadhi/web/Extensions.java:106]: done

File: server/src/main/java/com/flipkart/varadhi/config/ServerConfiguration.java (hunk 25 above)
Comment (aayustark007-fk) [reply server/src/main/java/com/flipkart/varadhi/config/ServerConfiguration.java:31]: server goes in with the default provider, without that there is no way to set the role assignments

Comment (aayustark007-fk) [reply server/src/main/java/com/flipkart/varadhi/config/ServerConfiguration.java:31]: so we will have to do the reverse, if default auth provider is enabled, server is enabled by default

File: server/src/test/java/com/flipkart/varadhi/auth/DefaultAuthorizationProviderTest.java (hunk 26 above)
Comment (aayustark007-fk) [reply server/src/test/java/com/flipkart/varadhi/auth/DefaultAuthorizationProviderTest.java:44]: done

**Comment by aayustark007-fk** [comment]
> The authz handlers are marked authenticated, but they dont have authz checks. Based on your description, anything pending there?

I have to add new resource type IAM_POLICY so that I can process authorization for set/get IAM policy api. This I plan to do in a follow-up PR

**Comment by gauravAshok** [comment]
So, what is pending in this PR?

**Comment by aayustark007-fk** [comment]
> So, what is pending in this PR?

I have pushed all the requested changes, this PR is ready to be merged.
One remaining point is on enabling permissions authorization on authZ handlers but I rather keep it as a separate PR as it will require me to introduce new elements to ResourceType and ResourceAction enums and their handling might warrant further discussion.

File: server/src/main/java/com/flipkart/varadhi/config/ServerConfiguration.java (hunk 25 above)
Comment (aayustark007-fk) [reply server/src/main/java/com/flipkart/varadhi/config/ServerConfiguration.java]: done

File: server/src/main/java/com/flipkart/varadhi/services/AuthZService.java [hunk 31]
```diff
@@ -0,0 +1,121 @@
...
+    }
+
+    public RoleBindingNode getIAMPolicy(ResourceType resourceType, String resourceId) {
+        return roleBindingMetaStore.getRoleBindingNode(resourceType, resourceId);
+    }
+
+    public RoleBindingNode setIAMPolicy(ResourceType resourceType, String resourceId, IAMPolicyRequest binding) {
+        RoleBindingNode node = createOrGetRoleBindingNode(resourceId, resourceType);
+        node.setRoleAssignment(binding.getSubject(), binding.getRoles());
+        return updateRoleBindingNode(node);
```
Comment (aayustark007-fk) [inline server/src/main/java/com/flipkart/varadhi/services/AuthZService.java:74]: Add task to consider cleaning up empty role binding nodes (not P0)

File: server/src/main/java/com/flipkart/varadhi/web/v1/authz/AuthZHandlers.java [hunk 32]
```diff
@@ -0,0 +1,151 @@
...
+    public static final String REQUEST_PATH_PARAM_RESOURCE = "resource";
+    public static final String REQUEST_PATH_PARAM_RESOURCE_TYPE = "resource_type";
+
+    private final AuthZService authZService;
+
+    public AuthZHandlers(AuthZService authZService) {
+        this.authZService = authZService;
+    }
+
+    private List<RouteDefinition> getDebugHandlers() {
```
Comment (aayustark007-fk) [inline server/src/main/java/com/flipkart/varadhi/web/v1/authz/AuthZHandlers.java:35]: remove the debug handlers

File: server/src/main/java/com/flipkart/varadhi/services/AuthZService.java [hunk 33]
```diff
@@ -0,0 +1,121 @@
...
+
+    public AuthZService(MetaStore metaStore, RoleBindingMetaStore roleBindingMetaStore) {
+        this.metaStore = metaStore;
+        this.roleBindingMetaStore = roleBindingMetaStore;
+    }
+
+    public RoleBindingNode createRoleBindingNode(String resourceId, ResourceType resourceType) {
+        if (!isResourceValid(resourceId, resourceType)) {
+            throw new IllegalArgumentException(String.format(
+                    "Invalid resource id(%s) for resource type(%s).",
```
Comment (aayustark007-fk) [inline server/src/main/java/com/flipkart/varadhi/services/AuthZService.java:27]: this is an internal construct, rather throw entity not found here

File: server/src/main/java/com/flipkart/varadhi/services/AuthZService.java [hunk 34]
```diff
@@ -0,0 +1,121 @@
...
+        RoleBindingNode node = new RoleBindingNode(resourceId, resourceType, new HashMap<>(), 0);
+        roleBindingMetaStore.createRoleBindingNode(node);
+        return node;
+    }
+
+    public List<RoleBindingNode> getAllRoleBindingNodes() {
+        return roleBindingMetaStore.getRoleBindingNodes();
+    }
+
+    public RoleBindingNode findRoleBindingNode(ResourceType resourceType, String resourceId) {
```
Comment (aayustark007-fk) [inline server/src/main/java/com/flipkart/varadhi/services/AuthZService.java:41]: find --> search that can return empty result
get --> exact lookup that can throw entity not found

File: server/src/main/java/com/flipkart/varadhi/services/AuthZService.java [hunk 35]
```diff
@@ -0,0 +1,121 @@
...
+        roleBindingMetaStore.deleteRoleBindingNode(resourceType, resourceId);
+    }
+
+    private RoleBindingNode createOrGetRoleBindingNode(String resourceId, ResourceType resourceType) {
+        boolean exists = roleBindingMetaStore.isRoleBindingPresent(resourceType, resourceId);
+        if (!exists) {
+            return createRoleBindingNode(resourceId, resourceType);
+        }
+        RoleBindingNode existingNode = roleBindingMetaStore.getRoleBindingNode(resourceType, resourceId);
+        if (existingNode.getResourceType() != resourceType) {
```
Comment (aayustark007-fk) [inline server/src/main/java/com/flipkart/varadhi/services/AuthZService.java:94]: this check is redundant

File: server/src/testE2E/java/com/flipkart/varadhi/DefaultAuthZProviderTests.java [hunk 36]
```diff
@@ -0,0 +1,327 @@
+package com.flipkart.varadhi;
```
Comment (aayustark007-fk) [inline server/src/testE2E/java/com/flipkart/varadhi/DefaultAuthZProviderTests.java:1]: add integration with existing org, team, etc crud flows (as other PR)
all varadhi apis are authorized if enabled

Comment (aayustark007-fk) [reply server/src/testE2E/java/com/flipkart/varadhi/DefaultAuthZProviderTests.java:1]: to allow authn to work either go the token way or impl default/fake authn provider to take identity from headers being passed.

**Review by aayustark007-fk — COMMENTED** [review]
need to do these changes as part of separate PR

**Review by kmrdhruv — APPROVED** [review]
LGTM,
As discussed, some pending refactoring will be taken as part of next PR.

File: server/src/main/java/com/flipkart/varadhi/services/AuthZService.java (hunk 31 above)
Comment (aayustark007-fk) [reply server/src/main/java/com/flipkart/varadhi/services/AuthZService.java:74]: done

File: server/src/main/java/com/flipkart/varadhi/auth/DefaultAuthorizationProvider.java [hunk 37]
```diff
@@ -0,0 +1,224 @@
...
+        }
+        List<Pair<ResourceType, ResourceContext>> leafToRootResourceIds =
+                generateResourceContextHierarchy(action, resource);
+        if (leafToRootResourceIds.isEmpty()) {
+            return Future.succeededFuture(false);
+        }
+
+        boolean result =
+                leafToRootResourceIds.stream()
+                        .anyMatch(pair -> isAuthorizedInternal(userContext.getSubject(), action, pair.getValue()));
```
[the remark this answers is not in the document]
Comment (aayustark007-fk) [reply server/src/main/java/com/flipkart/varadhi/auth/DefaultAuthorizationProvider.java:88]: done
